include/kwk/context/sycl/context.hpp

```diff
@@ -0,0 +1,441 @@
+    {
+      map(f, inout(c0), inout(cs)...);
+    }
+
+
+    template<typename Func>
+    void map_index(Func f, concepts::sycl::proxy auto&& p0)
+    {
+      parent::submit([&](::sycl::handler &h) 
+      {
+        auto acc = p0.access(h);
+        h.parallel_for(p0.size(), [=](auto i) { f(KWK_FWD(acc)[i], i.get_id(1)); });
+      });
+      parent::wait();
+    }
+
+
+    template<typename Func, concepts::container In, typename InProxy>
+    auto reduce_internal_v2_proxy ( In const& in
+                                  , InProxy& in_proxy
+                                  , [[maybe_unused]] Func func
+                                  , [[maybe_unused]] auto init
+                                  , std::size_t repeat_for_bench = 1
+                                  )
+    {
+      using data_type = typename In::value_type;
+      
+      data_type result = init;
+      {
+        std::size_t const numel = in.numel();
+        ::sycl::buffer<data_type> result_buf(&result, ::sycl::range<1>(1));
+
+        for (std::size_t r = 0; r < repeat_for_bench; ++r)
+        {
+          parent::submit([&](::sycl::handler& h)
+          {
+            ::sycl::accessor data_acc = in_proxy.access(h);
+            auto reduction_obj = ::sycl::reduction(result_buf, h, func);
+            
+              h.parallel_for( ::sycl::range<1>(numel)
+                            , reduction_obj
+                            , [=](::sycl::id<1> i, auto& sum)
+                              {
+                                sum.combine(data_acc[i]);
+                              }
+                            );
+          });
+          parent::wait();
+        }
+      }
+
+      return result;
+    }
+
+
+    template<typename Func, concepts::container In>
+    auto reduce_internal_v2(In const& in, [[maybe_unused]] Func func, [[maybe_unused]] auto init)
+    {
+      auto in_proxy = this->in(in);
+      return reduce_internal_v2_proxy(in, in_proxy, func, init);
+    }
+
+
+    template<typename Func, concepts::container In>
+    auto reduce([[maybe_unused]] In const& in, [[maybe_unused]] Func f, [[maybe_unused]] auto init)
+    {
+      try
+      {
+        return reduce_internal_v2(in, f, init);
+      }
+      catch (::sycl::exception const &e)
+      {
+        std::cout << "An exception is caught for SYCL reduce.\n";
+        std::terminate();
+      }
+    }
+
+    template<typename Func, concepts::container In, typename InProxy>
+    auto reduce_proxy ( [[maybe_unused]] In const& in
+                      , [[maybe_unused]] InProxy& in_proxy
+                      , [[maybe_unused]] Func f
+                      , [[maybe_unused]] auto init
+                      , std::size_t repeat_for_bench = 1
+                      )
+    {
+      try
+      {
+        return reduce_internal_v2_proxy(in, in_proxy, f, init, repeat_for_bench);
+      }
+      catch (::sycl::exception const &e)
+      {
+        std::cout << "An exception is caught for SYCL reduce.\n";
```

Comment:
Just rethrow the exception please, never terminate on the face of people

Reply:
Applied, thanks!